src/core/unit.c

```diff
@@ -3195,6 +3243,14 @@ int unit_add_dependency_by_name(Unit *u, UnitDependency d, const char *name, boo
         if (u->manager && FLAGS_SET(u->manager->test_run_flags, MANAGER_TEST_RUN_IGNORE_DEPENDENCIES))
                 return 0;
 
+        r = unit_is_recursive_template_dependency(u, name);
+        if (r < 0)
+                return r;
+        if (r == 1) {
+                log_unit_warning(u, "Dropping dependency %s that would lead to infinite recursion", name);
```

Comment:
Maybe change "would lead to" to "could lead to" or even "raises suspicion of infinite recursion"? By saying "would" we open ourselves to critique that we got the condition wrong, but also we could confuse the user if there is in fact no recursion.